File: dice.rb
# frozen_string_literal: true
module Irrgarten
  class Dice
    @@MAX_USES= 5;
    @@MAX_INTELLIGENCE= 10.0
    @@MAX_STRENGTH= 10.0
    @@RESURRECT_PROB= 0.3
    @@WEAPONS_REWARD= 2
    @@SHIELDS_REWARD= 3
    @@HEALTH_REWARD= 5
    @@MAX_ATTACK= 3
    @@MAX_SHIELD= 2

    @@generator= Random.new

    def random_pos (max)
      @@generator.rand(0...max)
    end

    def who_starts (nplayers)
      @@generator.rand(0...nplayers)
    end

    def random_intelligence
      @@generator.rand(0.0...@@MAX_INTELLIGENCE)
    end

    def random_strength
      @@generator.rand(0.0...@@MAX_STRENGTH)
    end

    def resurrect_player
      prob = @@generator.rand(0.0..1.0)
      prob > @@RESURRECT_PROB
      #return true if prob <= @RESURRECT_PROB
      #return false if prob > @RESURRECT_PROB
    end

    def weapon_reward
      @@generator.rand(0..@@WEAPONS_REWARD)
    end

    def shields_reward
      @@generator.rand(0..@@SHIELDS_REWARD)
    end

    def health_reward
      @@generator.rand(0..@@HEALTH_REWARD)
    end

    def weapon_power
      @@generator.rand(0...@MAX_ATTACK)
    end

    def shield_power
      @@generator.rand(0...@@MAX_SHIELD)
    end

    def uses_left
      @@generator.rand(0...@@MAX_USES)
    end

    def intensity (competence)
      @@generator.rand(0...competence)
    end

    def discard_element (uses_Left)

      return true if uses_Left == 0
      return false if uses_Left == @@MAX_USES

      probabylity = 1.0-(uses_Left.to_f/@@MAX_USES)
      @@generator.rand(0.0..1.0) >= probabylity
    end
  end
end

File: Dice.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;
import java.util.Random;
import java.util.ArrayList;
/**
 *
 * @author gonzaalovd & carlitros_gamer20
 */
public class Dice {
    static private final int MAX_USES = 5;
    static private final float MAX_INTELLIGENCE = (float) 10.0; //si da fallo hay que poner (float) o 10.0f
    static private final float MAX_STRENGTH = (float)10.0;
    static private final float RESURRECT_PROB = (float)0.3;
    static private final int WEAPONS_REWARD = 2;
    static private final int SHIELD_REWARD = 3;
    static private final int HEALTH_REWARD = 5;
    static private final int MAX_ATTACK = 3;
    static private final int MAX_SHIELD = 2;
    
    static private final Random generator = new Random(); //el new asigna una direccion de memoria a generator que se le pedirán
                                                         // valores aleatorios pero nunca cambia
    
    static public int randomPos(int max){
        int pos = generator.nextInt(max);
        return pos;
    }
    
    static public int whoStarts(int nplayers){
        int starts = generator.nextInt(nplayers);
        return starts;
    }
    
    static public float randomIntelligence(){
        return Dice.generator.nextFloat()*Dice.MAX_INTELLIGENCE;
        //float intel = generator.nextFloat()*MAX_INTELLIGENCE;
        //return intel;
    }
    
    static public float randomStrength(){
        return Dice.generator.nextFloat()*Dice.MAX_STRENGTH;
        //float str = generator.nextFloat()*MAX_STRENGTH;
        //return str;
    }
    
    static public boolean resurrectPlayer(){
        float prob = generator.nextFloat();
        return prob <= RESURRECT_PROB;
    }
    
    static public int weaponsReward(){
        return Dice.generator.nextInt(WEAPONS_REWARD+1);
        //int nweapons = generator.nextInt(WEAPONS_REWARD+1);
        //return nweapons;
    }
    
    static public int shieldsReward(){
        return Dice.generator.nextInt(Dice.SHIELD_REWARD+1);
        //int nshield = generator.nextInt(SHIELD_REWARD+1);
        //return nshield;
    }
    
    static public int healthReward(){
        return Dice.generator.nextInt(Dice.HEALTH_REWARD+1);
        //int nhealth = generator.nextInt(HEALTH_REWARD+1);
        //return nhealth;
    }
    
    static public float weaponPower(){
        float attack = generator.nextFloat()*MAX_ATTACK;
        
        while(attack == MAX_ATTACK){
        attack = generator.nextFloat()*MAX_ATTACK;
        }
        
        return attack;
    }
    
    static public float shieldPower(){
        float shield = generator.nextFloat()*MAX_ATTACK;
        
        while(shield == MAX_SHIELD){
            shield = generator.nextFloat()*MAX_SHIELD;
        }
        return shield;
    }
    
    static public int usesLeft(){
        return Dice.generator.nextInt(Dice.MAX_USES+1);
    }
    
    static public float intensity(float competence){
        return Dice.generator.nextFloat()*competence;
    }
    
    static public boolean discardElement(int usesLeft){
        float prob = 1-((float)usesLeft/MAX_USES);
        
        if(usesLeft == MAX_USES){
            return false;
        }
        else{
            if(usesLeft == 0){
                return true;
            }
            else{
                return prob >= Dice.generator.nextFloat();
            }
        }
    }
    
    static public Directions nextStep(Directions preference, ArrayList<Directions> validMoves, float intelligence){
        float prob = intelligence/MAX_INTELLIGENCE;
        
        if(Dice.generator.nextFloat() <= prob){
            return preference;
        }
        else{
            return validMoves.get(generator.nextInt(validMoves.size()));
        }
    }
}
